packages/protocol/contracts/layer2/shasta/impl/Anchor.sol

```diff
@@ -3,96 +3,224 @@ pragma solidity ^0.8.24;
+    // State Variables
+    // -------------------------------------------------------------------------
 
-    /// @dev Private storage for the current anchor state
+    /// @dev Current anchor state (4 storage slots)
     State private _state;
 
-    uint256[49] private __gap;
+    /// @dev Storage gap for future upgrades
+    uint256[46] private __gap;
 
     constructor(
         address _anchorTransactor,
+        uint32 _minGasIssuancePerSecond,
         IBondManager _bondManager,
-        IBlockHashManager _blockHashManager
+        IBlockHashManager _blockHashManager,
+        ISyncedBlockManager _syncedBlockManager
     )
         nonZeroAddr(_anchorTransactor)
-        nonZeroAddr(address(_bondManager))
-        nonZeroAddr(address(_blockHashManager))
         EssentialContract()
     {
-        bondManager = _bondManager;
         anchorTransactor = _anchorTransactor;
+        minGasIssuancePerSecond = _minGasIssuancePerSecond;
+        bondManager = _bondManager;
         blockHashManager = _blockHashManager;
+        syncedBlockManager = _syncedBlockManager;
     }
 
-    /// @notice Initialize the contract
-    /// @param _owner The owner of the contract
-    function init(address _owner) external initializer {
+    // -------------------------------------------------------------------------
+    // External Functions
+    // -------------------------------------------------------------------------
+
+    /// @notice Initializes the anchor contract with initial state
+    /// @param _owner Contract owner address for admin functions
+    /// @param _initialState Initial anchor state configuration
+    function init(address _owner, State memory _initialState) external initializer {
         __Essential_init(_owner);
+        _state = _initialState;
     }
 
-    /// @notice Retrieves the current anchor state
-    /// @return The current State struct containing synchronization data
+    /// @notice Returns the current anchor state
+    /// @return Current state including block info, gas issuance, and bond operations
     function getState() external view returns (State memory) {
         return _state;
     }
 
-    /// @notice Updates the anchor state with new values
-    /// @param _newState The new state to be set
-    /// @param _bondOperations The bond operations to be performed
-    /// @dev Only the anchor transactor address can call this function
-    ///      This ensures state updates come from the L2 system itself
+    /// @notice Atomically updates the anchor state with L1 synchronization data
+    /// @param _newState New state containing L1 block info and configuration
+    /// @param _bondOperations Bond credit operations to execute
+    /// @dev Security: Only callable by the keyless anchor transactor address
+    /// TODO(daniel): remove all validations as node cannot afford reverting.
     function setState(
         State memory _newState,
         LibBondOperation.BondOperation[] memory _bondOperations
```

Comment:
If we still want to keep the ethereum `block` structure as it is (which i think is important since otherwise it will break the hash calculation for all the ethereum tools), seems it's not easy to put `LibBondOperation.BondOperation[]` as a parameter of a system call (maybe still possible if we limit the length),  all parameters that used for a system call method should either be derived from the current chain state or in `block.extradata` (or some unused fields in block for L2), if not, the P2P sync can't work, they can't know how to recover the block without reading the L1 again...

Reply:
Damn, this is an issue then.....

Reply:
Rn i think maybe we can use `block. withdrawals`... let me also think about it...

Reply:
> Rn i think maybe we can use block.withdrawals... let me also think about it...

After double-checking, I think we can reuse it.